docs/readmes/proposals/p020_sgi_tunnel_transport.md

```diff
@@ -0,0 +1,114 @@
+This same tunnel can be potentially used for routing traffic to PGW aggregator host.
+
+## Design
+
+Initially IPSec tunnel was considered for tunneling traffic to the aggregator host. But after exploring other tunneling techniques it became clear that we need look at wireguard first. There are following advantages using wireguard over IPSec tunnels:
+
+1. wireguard performs better than IPsec or OpenVPN. There are bunch of benchmarking results on internet.
+2. It avoid dependency on OVS IPSec support.
+3. AGW can apply QoS policies before egressing traffic on the tunnel.
+
+So the plan is to implement wireguard based tunnel first and then implement IPSec. This doc covers both tunnel implementation.
+
+
+## **Detailed Design**
+
+This feature has two important aspects:
+
+1. Configuration of WireGuard tunnel
+2. Datapath programming
+
+### **Configuration:**
+
+AGW need following configuration for WireGuard tunnel:
+
+1. Install required binaries for WireGuard Tunnel
+2. Enable or disable SGi Tunnel
+3. Type of SGi Tunnel, wireguard, IPSec, etc
+4. Tunnel configuration:
+
+
+## **Datapath Programming for WireGuard Tunnel**
+
+![WireGuard datapath in NonNAT setup](assets/agw_wg_tun1.png)
+
+There are three parts in datapath programming.
+
+1. Setup the WG tunnel
+    1. Create the tunnel device
+    2. Apply the configuration from magma config
+        1. If the WG key is not provided generate the GW key
+        2. export the key over directoryD for operator visibility.
+    3. Tunnel local IP address, this would allow remote ssh/connectivity directly to the AGW from aggregator.
+    4. Store the configuration on persistent device.
+    5. Link the configuration to systemd to have the device configured on reboots.
+2. IP Address allocations for UE: WG tunnel transport would support existing IP allocation scheme
+    1. IP-POOL or Static IP based UE IP allocation would work the best, since there would be any need to configure uplink-bridge
+    2. In case DHCP based allocation is required we need to setup uplink-br0 for managing the DHCP traffic.
+3. Setup the flows to drive the traffic to/from the tunnel port.
+
+
+
+## **Implementation plan**
+
+1. This implementation would focus on introducing WG tunnel configuration via pipelined yml file. The WG tunnel would be implemented using WG commandline.
+2. PipelineD would create WG
+3. Use key configuration from magma configuration and map it to WG tunnel configuration
+4. Egress packet handling:
+    1. In case WG is enabled, push all uplink packets to the WG tunnel.
+5. Ingress packet:
+    1. ARP packet: send it to ARP responder table.
+    2. Regular IP packet: handle it as a downlink packet.
+
+## **Testing Plan**
+
+I am planning on writing integration tests to verify this in a vagrant development environment.
```

Comment:
Could you add a few more details on these tests

Reply:
done.